R/scalevariableset.R

```diff
@@ -106,3 +109,76 @@ removeSUMColumns <- function(df)
     }else  # NumberGrid
         return(df[, !grepl("^SUM, |, SUM$", colnames(df))])
 }
+
+#' Get numbers from attributtes of Pick One and Pick One - Multi questions
+#'
+#'  Extracts the codeframe and value attributes from the (R data.frame/factor representation of) Displayr's Categorical Variable Sets (PickOne and PickOneMulti questions in Q) and applies them to the input factors/data.frame to convert it to its underlying values.
+#' @param x A variable set from Q/Displayr (an R factor or data.frame
+#'     with special attributes).
+#' @return A matrix of numeric values with number of rows equal \code{NROW(x)}.
+#' @details levels(x) are always present in the names of the codeframe
+#'     attribute, \emph{unless} codes are hidden. The codeframe
+#'     contains sourcevalues for each code/level, which need to be
+#'     mapped from the sourcevalues attribute (original values
+#'     as read in with the data set) to the (possibly user-modified)
+#'     values attribute.
+#'
+#'     A level that is the result of a merge will contain multiple
+#'     source values for that element in the codeframe. These are
+#'     mapped to the corresponding underlying values in the values
+#'     attribute and averaged.
+#'
+#' @noRd
+numbersFromCategoricalVariableSets <- function(x)
+{
+    vv <- attr(x, "variablevalues")
+    if (is.list(vv))  # multi-variable variable set
+    {
+        if (length(vv) != ncol(x))  # probably not needed
+            stop("Invalid variable set provided; the length of the ",
+                 dQuote("variablevalues"), " attribute does not match ",
+                 " the number of variables in the variable set.")
+
+        ## out <- mapply(function(fact, vals) vals[levels(fact)[fact]], x, vv,
+        ##               SIMPLIFY = TRUE)
+        sv <- attr(x, "variablesourcevalues")
+        cf <- attr(x, "codeframe")
+
+        return(mapply(numbersFromCategoricalQVariable, x, vv, sv,
+                      MoreArgs = list(code.frame = cf),
+                      SIMPLIFY = TRUE))
```

Comment:
i.e. set `SIMPLIFY = FALSE` seems preferable?

Reply:
Because `numbersFromCategoricalQVariable` is our own function and we know/are guaranteed what it returns, I don't mind using `SIMPLIFY = TRUE` here. That behaviour is already replied on by `ScaleVariableSet` as well.